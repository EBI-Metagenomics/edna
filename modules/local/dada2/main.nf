
process DADA2 {
    // Run DADA2 pipeline including read-tracking
    tag "$meta.id"
    label "dada2_resources"
    container 'docker://quay.io/microbiome-informatics/dada2:v1'

    input:
    tuple val(meta), path(reads)

    output:
    tuple val(meta), path("*map.txt"), path("*asvs.fasta"), path("*_filt.fastq.gz"), optional: true, emit: dada2_out
    tuple val(meta), path("*_dada2_stats.tsv")                                     , optional: true, emit: dada2_stats
    tuple val(meta), path("*_dada2_errors.txt")                                    , optional: true, emit: dada2_errors
    tuple val(meta), env(stats_fail)                                               , optional: true, emit: dada2_stats_fail
    path "versions.yml"                                                            , emit: versions
    
    script:
    if ( meta.single_end ){
        """
        output_file="${meta.id}_dada2_output.txt"
        error_file="${meta.id}_dada2_errors.txt"
        dada2.R ${meta.id} $reads 2> \$error_file

        stats_fail=false
        if [[ -s \$error_file ]] && grep -q "Caught an error" \$error_file; then
            stats_fail=true
            mv \$output_file \$error_file
        fi

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            R: \$( R --version | head -1 | cut -d' ' -f3 )
            dada2: \$( R -e "suppressMessages(library(dada2));packageDescription('dada2')" | grep 'Version' | cut -d' ' -f2 )
        END_VERSIONS
        """
    } else {
        """
        output_file="${meta.id}_dada2_output.txt"
        error_file="${meta.id}_dada2_errors.txt"
        dada2.R ${meta.id} ${reads[0]} ${reads[1]} 2> \$output_file

        stats_fail=false
        if [[ -s \$output_file ]] && grep -q "Caught an error" \$output_file; then
            stats_fail=true
            mv \$output_file \$error_file
        fi

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            R: \$( R --version | head -1 | cut -d' ' -f3 )
            dada2: \$( R -e "suppressMessages(library(dada2));packageDescription('dada2')" | grep 'Version' | cut -d' ' -f2 )
        END_VERSIONS
        """
    }

}
