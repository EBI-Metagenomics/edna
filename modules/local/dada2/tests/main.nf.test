nextflow_process {

    name "Test Process DADA2"
    script "modules/local/dada2/main.nf"
    process "DADA2"
    tag "modules"
    tag "dada2"

    test("Should run without failures") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [ 
                        file('${projectDir}/modules/local/dada2/tests/data/test_noambig_1.fastq.gz', checkIfExists: true),
                        file('${projectDir}/modules/local/dada2/tests/data/test_noambig_2.fastq.gz', checkIfExists: true) 
                    ]
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.dada2_out).match() }
            )
        }

    }

}
