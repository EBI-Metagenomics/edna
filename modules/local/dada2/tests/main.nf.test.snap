{
    "Should run without failures": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    [
                        "test_1_map.txt:md5,95af731d6fa626206f9cc99e14f753b9",
                        "test_2_map.txt:md5,0843e375a960cc79e3a3f7ed3a3dda8e"
                    ],
                    "test_asvs.fasta:md5,c2bd49dd4d800233c146e03bf10e7250",
                    [
                        "test_1_filt.fastq.gz:md5,a6909a9f9d30ab033b3b1ff5456f4352",
                        "test_2_filt.fastq.gz:md5,8dff2a0f3928a5753b59f51a2ec27fe1"
                    ]
                ]
            ]
        ],
        "timestamp": "2024-05-08T15:16:27.863119401"
    }
}